### PR TITLE
feat: Simple mxnet example for MNIST

### PR DESCRIPTION
**Pull request by Rishav1**
Addresses #350

### CODE REVIEW

File: examples/mxnet_simple.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    model.fit(X=train, eval_data=val, batch_end_callback=_callback)
+
+    # Return the 1.0 - accuracy
+    test = mx.io.NDArrayIter(
+        data=mnist['test_data'],
+        label=mnist['test_label'],
+        batch_size=BATCHSIZE)
+    accuracy = model.score(X=test)
+
+    return 1.0 - accuracy
```
Comment (sile) [inline examples/mxnet_simple.py]: Since v0.9.0, Optuna supports optimizations in the direction of maximization (see https://github.com/pfnet/optuna/pull/331 for details).
Therefore, you can return `accuracy` directly if `study` instance has created with `direction='maximize'` option.

File: examples/mxnet_simple.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+"""
```
Comment (sile) [inline examples/mxnet_simple.py]: Please add this example to the example list in [examples/README.md](https://github.com/pfnet/optuna/blob/master/examples/README.md#examples-with-ml-libraries).

File: examples/mxnet_simple.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    print('Best trial:')
+    trial = study.best_trial
+
+    print('  Value: ', trial.value)
+
+    print('  Params: ')
+    for key, value in trial.params.items():
+        print('    {}: {}'.format(key, value))
+
+    print('  User attrs:')
```
Comment (sile) [inline examples/mxnet_simple.py]: I think this and the following `print` calls are unnecessary because user attributes are never set in this example.

File: examples/mxnet_simple.py (hunk 2 above)
Comment (sile) [reply examples/mxnet_simple.py]: In addition, please add `mxnet` as a package that installed when the daily CI is executed (see: [.circleci/config.yml](https://github.com/pfnet/optuna/blob/master/.circleci/config.yml#L191)).

File: examples/mxnet_simple.py (hunk 1 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Got it.

File: examples/mxnet_simple.py (hunk 2 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Okay.

File: examples/mxnet_simple.py (hunk 3 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Removed it.

File: examples/mxnet_simple.py (hunk 2 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Done.

File: examples/mxnet_simple.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+
+N_TRAIN_EXAMPLES = 3000
+N_TEST_EXAMPLES = 1000
+BATCHSIZE = 128
+EPOCH = 10
+DIR = tempfile.mkdtemp()
+
+
+def get_mnist():
```
Comment (sile) [inline examples/mxnet_simple.py]: Is this function needed?

I prefer the previous version that simply calls the predefined `mx.test_utils.get_mnist()` for getting MNIST dataset (simplicity is very important to example code).
Did you have any troubles with `mx.test_utils.get_mnist()`.

Comment (Rishav1) [reply examples/mxnet_simple.py]: mx.test_utils.get_mnist() was downloading the data in the Optuna folder and unzipping it there itself. Since it doesn't take any parameters, there was no proper way of stopping that from happening besides the one I took.

Comment (sile) [reply examples/mxnet_simple.py]: Thank you for your explanation. I was able to understand the motivation to define the `get_mint` function.
Although the behaviour of `mx.test_utils.get_mnist()` feels somewhat annoying, I think it is acceptable for this case (i.e., only used in the example code).
As I said in the first comment, I would like to make examples as concise and easy to understand as possible.

Comment (Rishav1) [reply examples/mxnet_simple.py]: Okay. Got it.

File: examples/mxnet_simple.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Trainer
+    model = mx.model.FeedForward(
+        symbol=mlp,
+        optimizer=optimizer,
+        num_epoch=EPOCH,
+    )
+    model.fit(X=train, eval_data=val, batch_end_callback=_callback)
+
+    # Return the 1.0 - accuracy
```
Comment (sile) [inline examples/mxnet_simple.py]: This comment should be updated.

File: examples/README.md [hunk 6]
```diff
@@ -13,6 +13,7 @@ This page contains a list of example codes written with Optuna.
 * [Chainer](./chainer_simple.py)
 * [ChainerMN](./chainermn_simple.py)
 * [LighGBM](./lightgbm_simple.py)
+* [Mxnet](./mxnet_simple.py)
```
Comment (sile) [inline examples/README.md]: Please replace "Mxnet" by "MXNet" (the latter is the official spelling).

File: examples/mxnet_simple.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    # Trainer
+    model = mx.model.FeedForward(
+        symbol=mlp,
+        optimizer=optimizer,
+        num_epoch=EPOCH,
+    )
+    model.fit(X=train, eval_data=val, batch_end_callback=_callback)
+
+    # Return the  accuracy
```
Comment (sile) [inline examples/mxnet_simple.py]: This sentence contains extra whitespace, so please delete it.

Comment (sile) [reply examples/mxnet_simple.py]: Besides, please add a trailing period (cf. https://github.com/pfnet/optuna/pull/364#discussion_r271557606).

File: examples/mxnet_simple.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
+"""
+Optuna example that optimizes multi-layer perceptrons using MX-NET.
```
Comment (toshihikoyanase) [inline examples/mxnet_simple.py]: Please replace 'MX-NET' with 'MXNet' as can be seen in [the official site](https://mxnet.apache.org/).

File: examples/mxnet_simple.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+In this example, we optimize the validation accuracy of hand-written digit recognition using
+MX-NET and MNIST. We optimize the neural network architecture as well as the optimizer
```
Comment (toshihikoyanase) [inline examples/mxnet_simple.py]: ditto.

File: examples/mxnet_simple.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        if param.eval_metric is not None:
+            if param.epoch == 0 and param.nbatch == 1:
+                print("%-12s %-12s" % ("epoch", "val-accuracy"))
+            name_value = param.eval_metric.get_name_value()
+            if param.nbatch == 1:
+                for name, value in name_value:
+                    print('%-12d %-12f' % (param.epoch, value))
+
+    # Trainer
+    model = mx.model.FeedForward(
```
Comment (toshihikoyanase) [inline examples/mxnet_simple.py]: I encountered the following warning here.
> DeprecationWarning: mxnet.model.FeedForward has been deprecated. Please use mxnet.mod.Module instead.

Please replace `mx.model.FeedForward` with `mx.mod.Module`.

File: examples/mxnet_simple.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        symbol=mlp,
+        optimizer=optimizer,
+        num_epoch=EPOCH,
+    )
+    model.fit(X=train, eval_data=val, batch_end_callback=_callback)
```
Comment (toshihikoyanase) [inline examples/mxnet_simple.py]: I briefly examined the way to show metrics log in mxnet applications and found that they usually used the root logger. To show the metrics, we just set the logging level as shown in [the official example of mxnet](https://github.com/apache/incubator-mxnet/blob/master/example/svm_mnist/svm_mnist.py#L34), and we do not need to define the custom callback function.

The log messages are more crowded than those generated by `_callback` as follows, but I think we can use the way that is familiar with mxnet users.

> INFO:root:Epoch[0] Train-accuracy=0.108724
> INFO:root:Epoch[0] Time cost=0.087
> INFO:root:Epoch[0] Validation-accuracy=0.129883

What do you think of it?

File: examples/mxnet_simple.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+import numpy as np
+
+
+N_TRAIN_EXAMPLES = 3000
+N_TEST_EXAMPLES = 1000
+BATCHSIZE = 128
+EPOCH = 10
+
+
+def model_fn(trial):
```
Comment (toshihikoyanase) [inline examples/mxnet_simple.py]: Could you change the function name to `create_model` because it seems to be more consistent with `create_optimizer` than `model_fn`.

File: examples/mxnet_simple.py (hunk 11 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: I get it. removing the callback is a good call.

File: examples/mxnet_simple.py (hunk 12 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Yeah. Will do.

File: examples/mxnet_simple.py (hunk 10 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Okay. Got it.

File: examples/mxnet_simple.py (hunk 9 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Alright.

File: examples/mxnet_simple.py (hunk 8 above)
Comment (Rishav1) [reply examples/mxnet_simple.py]: Yeah. I missed it I guess.

File: examples/README.md (hunk 6 above)
Comment (Rishav1) [reply examples/README.md]: Okay.

**Comment by sile** [comment]
🎉
